codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java

```diff
@@ -0,0 +1,297 @@
+    private final Http2Connection.PropertyKey stateKey;
+    private final State connectionState;
+    /**
+     * The minimum number of bytes that we will attempt to allocate to a stream. This is to
+     * help improve goodput on a per-stream basis.
+     */
+    private int allocationQuantum = 1024;
+
+    public WeightedFairQueueByteDistributor(Http2Connection connection) {
+        stateKey = connection.newKey();
+        Http2Stream connectionStream = connection.connectionStream();
+        connectionStream.setProperty(stateKey, connectionState = new State(connectionStream, 16, 0));
+
+        // Register for notification of new streams.
+        connection.addListener(new Http2ConnectionAdapter() {
+            @Override
+            public void onStreamAdded(Http2Stream stream) {
+                stream.setProperty(stateKey, new State(stream));
+            }
+
+            @Override
+            public void onWeightChanged(Http2Stream stream, short oldWeight) {
+                Http2Stream parent;
+                if (state(stream).isActiveCountForTree != 0 && (parent = stream.parent()) != null) {
+                    state(parent).totalQueuedWeights += stream.weight() - oldWeight;
+                }
+            }
+
+            @Override
+            public void onStreamClosed(Http2Stream stream) {
+                state(stream).close();
+            }
+
+            @Override
+            public void onPriorityTreeParentChanged(Http2Stream stream, Http2Stream oldParent) {
+                Http2Stream parent = stream.parent();
+                if (parent != null) {
+                    State state = state(stream);
+                    if (state.isActiveCountForTree != 0) {
+                        State pState = state(parent);
+                        pState.offer(state);
+                        pState.isActiveCountChangeForTree(state.isActiveCountForTree);
+                    }
+                }
+            }
+
+            @Override
+            public void onPriorityTreeParentChanging(Http2Stream stream, Http2Stream newParent) {
+                Http2Stream parent = stream.parent();
+                if (parent != null) {
+                    State state = state(stream);
+                    if (state.isActiveCountForTree != 0) {
+                        State pState = state(parent);
+                        pState.remove(state);
+                        pState.isActiveCountChangeForTree(-state.isActiveCountForTree);
+                    }
+                }
+            }
+        });
+    }
+
+    @Override
+    public void updateStreamableBytes(StreamState state) {
+        state(state.stream()).updateStreamableBytes(state.streamableBytes(),
+                                                    state.hasFrame() && state.isWriteAllowed());
+    }
+
+    @Override
+    public boolean distribute(int maxBytes, Writer writer) throws Http2Exception {
```

Comment:
That sounds fine to me, and it seems to handle all the cases. Alternatives seem much more complex and not necessarily worth it. Even if we choose something else later, that seems like a good solution at this point.